File: karatsuba/karatsuba.py
def karatsuba(x,y):
    n = max(len("{0:b}".format(x)), len("{0:b}".format(y)))
    if (n <= 10):
        return x * y
    else:
        n = (n / 2) + (n % 2)
        b = x >> n
        a = x - (b << n)
        d = y >> n
        c = y - (d << n)
        
        ac = karatsuba(a, c)
        bd = karatsuba(b, d)
        abcd = karatsuba(a + b, c + d)
        return ac + (bd << 2 * n) + ((abcd - ac - bd) << n)

print(karatsuba(3141592653589793238462643383279502884197169399375105820974944592, 2718281828459045235360287471352662497757247093699959574966967627))

File: karatsuba/karatsuba.cs
using System;
using System.Numerics;

namespace csharp
{
    static class Karatsuba
    {
        public static BigInteger multiply(BigInteger x, BigInteger y)
        {
            int n = Math.Max(countBits(x), countBits(y));
            if (n <= 10)
            {
                return x * y;
            }
            else
            {
                n = (n / 2) + (n % 2);
                BigInteger b = x >> n;
                BigInteger a = x - (b << n);
                BigInteger d = y >> n;
                BigInteger c = y - (d << n);

                BigInteger ac = multiply(a, c);
                BigInteger bd = multiply(b, d);
                BigInteger abcd = multiply(a + b, c + d);
                BigInteger x1 = (bd << 2 * n);
                BigInteger x2 = ((abcd - ac - bd) << n);
                return ac + x1 + x2;
            }
        }
        public static int countBits(BigInteger number)
        {
            return (int)Math.Log((double)number, 2.0) + 1;
        }
    }
}
